### PR TITLE
Issue stratisd 2134 setup pools prototype

### PR DESCRIPTION
**Pull request by jbaublitz**
Related to #2134 

@mulkieran Here is the stratisd-min prototype if you'd like to look it over.

### CODE REVIEW

**Review by mulkieran — CHANGES_REQUESTED** [review]
I took just a quick high-level look. I'll be looking more at the rpc service implementation later. So far it all makes sense, although I'm not familiar w/ the dracut-specific file to have much to say abou that. I have a couple of questions.

* Are the two files in the daemon subdirectory expected to be removed soon, to be replaced w/ rpc based code?
* Why remove the Rc from the Name type? I think the original purpose was to allow the inner value to be shared when using the Name as a key in the Table struct.

File: Makefile [hunk 1]
```diff
@@ -142,6 +152,7 @@ clippy:
 	license
 	outdated
 	release
+	stratis-min
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply Makefile:155]: Yes, you're right.

**Comment by jbaublitz** [comment]
> So far it all makes sense, although I'm not familiar w/ the dracut-specific file to have much to say abou that.

That needs to change once I write a unit file for stratisd-min. I would recommend against taking a look at that for now as that's the last step before testing it in the initramfs.

> Are the two files in the daemon subdirectory expected to be removed soon, to be replaced w/ rpc based code?

Yes, there is a lot of code clean up that I still have to do, so it might be best not to review this at all and merely take a look at the approach for now.

> Why remove the Rc from the Name type? I think the original purpose was to allow the inner value to be shared when using the Name as a key in the Table struct.

This caused the `StratEngine` type to be unsafe across threads. `Rc` cannot be passed between threads so the compiler was failing due to that wrapper inside the `Name` type. I can try adding it back in now that the architecture has changed slightly from `jsonrpc` to `jsonrpsee`. However, we may want to consider the alternative, `Arc`, which is thread-safe. Given that we intend to eventually add multi-threading in stratisd, my guess is that we will eventually bump into problems using thread-unsafe types when we add in tokio support. The interesting thing, however, is that if that `Rc` were necessary, I would expect a compiler error from the borrow checker and it doesn't appear that our code actually needs it right now.

File: Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (jbaublitz) [reply Makefile:155]: The target used to be `stratis-min` but now that we have stratisd-min too, I forgot to update the target to the new value, `build-min`.

**Comment by mulkieran** [comment]
> > Why remove the Rc from the Name type? I think the original purpose was to allow the inner value to be shared when using the Name as a key in the Table struct.
> 
> This caused the `StratEngine` type to be unsafe across threads. `Rc` cannot be passed between threads so the compiler was failing due to that wrapper inside the `Name` type. I can try adding it back in now that the architecture has changed slightly from `jsonrpc` to `jsonrpsee`. However, we may want to consider the alternative, `Arc`, which is thread-safe. Given that we intend to eventually add multi-threading in stratisd, my guess is that we will eventually bump into problems using thread-unsafe types when we add in tokio support. The interesting thing, however, is that if that `Rc` were necessary, I would expect a compiler error from the borrow checker and it doesn't appear that our code actually needs it right now.

If you look at the insert() method in Table, you'll see that it clones the name parameter twice. If you remove the Rc, then you get two copies of the same string in the Table structure instead of just two copies of the pointer. I think that was the whole point of introducing the Rc. I can't induce a compiler error by adding in the Rc's now.

**Comment by jbaublitz** [comment]
@mulkieran 

> > > Why remove the Rc from the Name type? I think the original purpose was to allow the inner value to be shared when using the Name as a key in the Table struct.
> > 
> > 
> > This caused the `StratEngine` type to be unsafe across threads. `Rc` cannot be passed between threads so the compiler was failing due to that wrapper inside the `Name` type. I can try adding it back in now that the architecture has changed slightly from `jsonrpc` to `jsonrpsee`. However, we may want to consider the alternative, `Arc`, which is thread-safe. Given that we intend to eventually add multi-threading in stratisd, my guess is that we will eventually bump into problems using thread-unsafe types when we add in tokio support. The interesting thing, however, is that if that `Rc` were necessary, I would expect a compiler error from the borrow checker and it doesn't appear that our code actually needs it right now.
> 
> If you look at the insert() method in Table, you'll see that it clones the name parameter twice. If you remove the Rc, then you get two copies of the same string in the Table structure instead of just two copies of the pointer. I think that was the whole point of introducing the Rc. I can't induce a compiler error by adding in the Rc's now.

Okay that makes sense! In that case, I'd recommend using `Arc` in preparation for multithreading. The issue is that the following code:

```
use std::{rc::Rc, sync::Mutex};  
                               
fn test_fn<T>(_t: T) where T: Send + Sync {
    
} 

fn main() {
    let test = Mutex::new(Rc::new(5));
    test_fn(test);             
} 
```
causes the following error:

```
error[E0277]: `std::rc::Rc<{integer}>` cannot be sent between threads safely
 --> test.rs:9:13                                          
  |                                                     
3 | fn test_fn<T>(_t: T) where T: Send + Sync {            
  |                               ---- required by this bound in `test_fn`
...
9 |     test_fn(test);
  |             ^^^^ `std::rc::Rc<{integer}>` cannot be sent between threads safely
  |
  = help: the trait `std::marker::Send` is not implemented for `std::rc::Rc<{integer}>`
  = note: required because of the requirements on the impl of `std::marker::Send` for `std::sync::Mutex<std::rc::Rc<{integer}>>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

while

```
use std::sync::{Arc, Mutex};  
                               
fn test_fn<T>(_t: T) where T: Send + Sync {
    
} 

fn main() {
    let test = Mutex::new(Arc::new(5));
    test_fn(test);             
} 
```

compiles just fine.

This is a minimal example that shows that `StratEngine` will not be `Send + Sync` with an `Rc` anywhere in its type tree. As you can see, even wrapping `Rc` in a synchronization primitive does not cause the whole data structure to be `Send + Sync` which we will need if passing something like `Arc<RwLock<StratEngine>>` across threads which is typically required for any kind of multithreading with tokio.

Also even `Arc<Mutex<Rc<u16>>>` fails to make it map a little bit more closely to the data structure that I'm planning to use.

What are your thoughts?

**Comment by jbaublitz** [comment]
Closing in favor of a new PR.

**Pull request closed without merging**